go/vt/vtgate/scatter_conn.go

```diff
@@ -386,6 +392,15 @@ func (stc *ScatterConn) Commit(context context.Context, session *SafeSession) (e
 
 // Rollback rolls back the current transaction. There are no retries on this operation.
 func (stc *ScatterConn) Rollback(context context.Context, session *SafeSession) (err error) {
+	if session == nil {
+		return fmt.Errorf("cannot rollback: empty session")
+	}
+	if !session.InTransaction() {
+		return fmt.Errorf("cannot rollback: not in transaction")
+	}
+	if len(session.ShardSessions) == 0 {
+		return fmt.Errorf("cannot rollback: invalid session, always update session from RPC response")
```

Comment:
I'm not sure if this check is valid. Do we need it? - it would prevent a client from issuing BEGIN followed by a ROLLBACK. This is what's causing the Java test failure.